### ISSUE TITLE
V2 Bug: _msgSender() can be spoofed

### ISSUE THREAD

**Issue by Boffee** [latticexyz/mud#601]
https://github.com/latticexyz/mud/blob/0ac76fd57484f54860157b79678b8b9eb7a86997/packages/world/src/WorldContext.sol#L10-L16

The sender can spoof arbitrary `_msgSender()` by manually setting the last 20 bytes of the `calldata`. `_msgSender()` should use `calldata` when `msg.sender` is the `World` contract (or a list of approved addresses) and `msg.sender` otherwise.

From a glance, this seems to mean that `WorldContext` needs to store the `World` address.

**Comment by holic** [comment]
ooc why was this closed?

**Comment by dk1a** [comment]
for context the reason this isn't an issue is that systems are always called by `World`, which always passes the correct sender

**Comment by Boffee** [comment]
Realized that `StoreSwitch` selects either the caller (`msg.sender`) or the contract itself (`address(this)`) as the Store, which means that a spoofed `_msgSender()` would not give the caller any additional write access.

**Comment by Boffee** [comment]
Actually, might want to warn somewhere that systems should only use `_msgSender()` when writing to MUD Stores. 

If a system calls another contract with `_msgSender()` as an argument or has states that depend on `_msgSender()` (for whatever reason), the caller would be able to change states that it doesn't have access to.

**Comment by Boffee** [comment]
Can someone else double check this? I'm confusing myself.

I just noticed that the autogenerated table libraries have a `function set(IStore _store, ...)`, which means the tables a contract can write to is not constrained to the caller and itself by design. 

If the intention is constrain systems to only have write access to the caller and itself, the caller (`World`) would need something like `onlyReentrant` which caches the system it calls before the call and check that the reentrant caller match the cached address.

**Comment by alvrs** [comment]
If you're using the World framework, there are two possible ways to write to the World store:
1. The System is delegatecalled from the World -> `StoreSwitch` writes directly to the World storage (since the world storage is delegated)
2. The system is called from the World -> `StoreSwitch` calls the `setRecord`/`setField` methods on the World contract (= `msg.sender` in this case), and the World performs access control checks on the write.

It's true that anyone can call a System and pass anything as the calldata accessed by `_msgSender()`, but since **writes always happen on the caller's storage** (see two cases above) it's not an issue. 


**However** this is only true if the System doesn't have any own internal state (which is strongly recommended). Unfortunately https://github.com/latticexyz/mud/pull/568 introduced a kind of internal storage - if a system is called with `msg.value`, the value is forwarded to the System. Now the System has to be extremely careful what to do with the value, it should probably not store it or allow access via `_msgSender()`. I think we should remove this behaviour of forwarding `msg.value` to Systems, and instead store all value on the World contract, which keeps track of value per namespace and performs access control checks like with storage writes.

**Comment by Boffee** [comment]
Hmm, I think I'm missing something here. Looking at the autogenerated table libraries, a system can target any store. Of course this would require store address to be specified by the caller or stored in the system contract as a state or in bytecode.

https://github.com/latticexyz/v2sandbox/blob/main/packages/contracts/src/tables/CounterTable.sol#L95-L100

From what I understand, "writes always happen on the caller's storage" is suggested by a standard way to implement systems, rather than enforced. Only the World contract can enforce that the call stack is `world -> system -> world`.

**Comment by alvrs** [comment]
> Looking at the autogenerated table libraries, a system can target any store. 

This is true, but the targeted Store contract is expected to perform access control checks on the caller, like the World contract does here:

https://github.com/latticexyz/mud/blob/cd224a5244ee55316d4b95a21007a8076adefe6e/packages/world/src/World.sol#L131-L147

> From what I understand, "writes always happen on the caller's storage" is suggested by a standard way to implement systems, rather than enforced

Yes, "writes always happen on the caller's storage" refers to the reference implementation - it's totally possible to explicitly call another Store contract that's not the caller, but it's not possible to avoid the access control checks.

For reference - the autogenerated table libraries use `StoreSwitch` to determine whether to call an (access controlled) method on the `msg.sender` or to use the libraries directly (to write to their own storage, which refers to the delegated storage of the caller in case of `delegatecall`): https://github.com/latticexyz/mud/blob/cd224a5244ee55316d4b95a21007a8076adefe6e/packages/store/src/StoreSwitch.sol#L81-L87

**Comment by Boffee** [comment]
> it's totally possible to explicitly call another Store contract that's not the caller, but it's not possible to avoid the access control checks.

`AccessControl.requireAccess` checks that the system address (`msg.sender`), but does not seem to check who called the system (can be the world contract, EOA, or any arbitrary contract).

This is not a problem as long as systems only uses `StoreSwitch` to access stores. However, developers might skip `StoreSwitch` to try to be smart with anti-patterns to do things like multi-store access, bypass world proxy, gas optimizations, code sharing etc if they're unaware of the limitations of `_msgSender()`. Additionally, this is deceptive because the autogenerated libraries provides the interface to easily do this.

**Comment by dk1a** [comment]
> AccessControl.requireAccess checks the system address (msg.sender), but does not seem to check who called the system (can be the world contract, EOA, or any arbitrary contract).

Calling the system has its own access check
https://github.com/latticexyz/mud/blob/cd224a5244ee55316d4b95a21007a8076adefe6e/packages/world/src/World.sol#L388-L389

You can make a public system, but it's expected to take care of its own access control properly then.

If you make a public system with bad access control, since each table also has access control, you can only break your own tables (or ones that you were given access to)

> This is not a problem as long as systems only use StoreSwitch to access stores

Well yeah if you use MUD you shouldn't try to skip StoreSwitch, do arbitrary calls, or use `storage`. a dedicated linter could help here

> Additionally, this is deceptive because the autogenerated libraries provide the interface to easily do this.

I don't really like that as the default actually, for now it's a workaround for tests and scripts (because they're not called by world)

**Comment by Boffee** [comment]
>> AccessControl.requireAccess checks the system address (msg.sender), but does not seem to check who called the system (can be the world contract, EOA, or any arbitrary contract).
>
> Calling the system has its own access check

Sorry, I meant specifically system which skips `StoreSwitch`. In this case, an exploiter would call the system contract directly instead of through the world contract, so the `publicAccess` check would be skipped. This applies to both public and private systems.

> Well yeah if you use MUD you shouldn't try to skip StoreSwitch, do arbitrary calls, or use msg.sender or storage.

My main point is that this should be emphasize this in documentation. Specifically, `_msgSender()` in `WorldContext` **can only be used when accessing the caller's (`msg.sender`) store/states** because it doesn't check for trusted forwarder like `ERC2771`.`_msgSender()` should not be used to read from or write to:

1. the contract's own store if it has one (even if the contract is the caller)
1. any other store that is not the caller
2. the state of any contract that is not the caller (technically, this is a superset of 1.)

Also, the necessity to use `StoreSwitch` with MUD was not obvious to me. (Might've been because I only skimmed the docs)

> a dedicated linter could help here

This could be helpful. Is there way to cover all possible implementations or would it be more for the average developer?

It's also possible to enforce `StoreSwitch` logic in the `World` contract by tracking the call stack. The main disadvantage is gas cost from temporarily caching the call stack in storage, which seems wasteful just to prevent developers from shooting themselves in the foot. Maybe do this for testing only?

In either case, should still warn that `_msgSender()` have different security properties from what people expects. Could possibly rename `_msgSender()` to something else?

**Comment by alvrs** [comment]
> AccessControl.requireAccess checks the system address (msg.sender), but does not seem to check who called the system (can be the world contract, EOA, or any arbitrary contract).

This is intended - Systems are the public entry point into the app and need to perform their own access control checks to prevent unauthorized access (just like a regular non-MUD contract has to have checks in place for methods modifying internal storage). From the perspective of the World contract these fine grained access control checks are irrelevant, it only needs to know that it is being modified by a System that is authorized to do so.

You're raising a fair point though: with the autogen method that allows passing an explicit Store, it's easy to shoot yourself in the foot because now the _msgSender() can not be trusted anymore. I agree with @dk1a that a MUD linter would be helpful here, and maybe we can find a way to run tests without requiring to pass the World as an explicit argument (so we can remove this function from autogen)

**Comment by Boffee** [comment]
>> AccessControl.requireAccess checks the system address (msg.sender), but does not seem to check who called the system (can be the world contract, EOA, or any arbitrary contract).
>
> This is intended - Systems are the public entry point into the app and need to perform their own access control checks to prevent unauthorized access 

Sorry, I'm mixing up terminologies. This is referring to `AccessControl.requireAccess` not being able to distinguish between:

a. `client -> world -> system -> world`
b. `client -> system -> world`

It is safe to trust the system's `_msgSender()` in (a) when used to access the world's store, but never safe in (b). A system that only uses `StoreSwitch` only allows (a), whereas a system that skips `StoreSwitch` also allows (b).

**Comment by alvrs** [comment]
> Sorry, I'm mixing up terminologies. This is referring to AccessControl.requireAccess not being able to distinguish between:
>
> a. client -> world -> system -> world
> b. client -> system -> world
>
> It is safe to trust the system's _msgSender() in (a) when it's used to access the world's store, but never safe in (b). A system that only uses StoreSwitch only allows (a), whereas a system that skips StoreSwitch also allows (b).

This is correct. For b) to work your system has to hardcode which store to set values on though, which is explicitly not recommended (and we should have a linter to prevent this if you want to use the World framework). It is still possible to use approach b), but in that case access control in the system can't rely on `_msgSender` (but needs to use `msg.sender` instead), and it won't be possible to use this System within the World framework anymore. Arguably it's not even a "System" but just a custom contract then.

In other words: approach a) and b) are mutually exclusive, approach a) is supported by the World framework, approach b) is not. If developers use the World framework + StoreSwitch, there is no security risk of `msgSender()` being spoofable.

**Comment by Boffee** [comment]
> For b) to work your system has to hardcode which store to set values on though

A system can also get the store address as a function argument, from calldata, by reading from `msg.sender`, or by reading from `_msgSender()`.

> If developers use the World framework + StoreSwitch, there is no security risk of msgSender() being spoofable

I agree. It should be emphasized that `_msgSender()` has no security risk only when it is used to write to the caller's store/states; using `_msgSender()` to write to the store/states of other contracts is not safe even when using the World framework + `StoreSwitch`.

I think one implication is that it might not possible to securely implement system hooks that both depend on `_msgSender()` and modify states. For example, if we use a system hook to trigger `Transfer` events on a ERC20 proxy, the system's `msg.sender` can spoof both its own store and `_msgSender()` at the same to emit a `Transfer` event without updating the balance in the real store. If this is the case, would it even be possible to implement token standards with MUD2?

**Comment by Boffee** [comment]
I hope I'm completely wrong about this, but a more general implication seem to be that non-root systems in apps built using the World framework + `StoreSwitch` should never be granted access to any contracts outside of the world (the world can still grant access to external contracts). This would mean that non-root systems can't compose on other stateful contracts (can't build dex, marketplaces, interoperable game assets etc.).

**Comment by alvrs** [comment]
> A system can also get the store address as a function argument, from calldata, by reading from msg.sender, or by reading from _msgSender().

Of course, "hardcoded" was the wrong word. What I mean is explicit logic that determines on which Store to write other than `StoreSwitch` (which always writes on the caller).

> I think one implication is that it might not possible to securely implement system hooks that both depend on _msgSender() and modify states. For example, if we use a system hook to trigger Transfer events on a ERC20 proxy, the system's msg.sender can spoof both its own store and _msgSender() at the same to emit a Transfer event without updating the balance in the real store. If this is the case, would it even be possible to implement token standards with MUD2 without hardcoding the world address or restricting hooks to root systems?

If you implement your Systems with the `StoreSwitch` method, only calls that come from the "real" World would update the "real" user balance. Anyone else can also call the System, but the System is only an implementation that will act on the caller's storage. For your ERC20 example, the ERC20 proxy would be a non-system contract, which just calls the corresponding methods on the World contract, which then calls the System implementation, which performs its checks using the World contracts storage. If the transfer succeeds, a System hook can trigger your ERC20 proxy to emit the correct event.

> I hope I'm completely wrong about this, but a more general implication seem to be that non-root systems in apps built using the World framework + StoreSwitch should never be granted access to any contracts outside of the world (the world can still grant access to external contracts). This would mean that non-root systems can't compose on other stateful contracts (can't build dex, marketplaces, interoperable game assets etc.).

You have multiple options:
1) You implement your Systems without internal state, which allows the System to be used as pure logic in different Worlds.
2) You limit your System to a single World (eg. by storing the World address in a immutable state variable) and then can use the System as a standalone contract (eg. to grant it access in another contract) in addition to via the World contract.
3) Instead of limiting your System to a single World, you can always use the "Proxy/Hook" pattern described above to create an individual contract that interfaces with a World

**Comment by Boffee** [comment]
Thanks a lot for the explanation. I misunderstood mud2 as a replacement for diamond pattern and mud1.

It made a lot more sense to me after thinking about mud2 as a sql database. The World is the access control, the Store is the storage, and the Systems are sql functions. Like sql functions, systems are stateless and should only write to the World/Store that calls them; **mud2 systems cannot (securely) write to external resources unlike mud1 systems and diamond pattern facets** (with the exception of root system which work the same way as facets).

If a mud app needs to write to external resources, it should do so through an external contracts which handles the external writes and forward the remaining world specific writes to the world. 

To implement staking in mud2, we would do something like:

```mermaid
flowchart TB;
  C[client] -."stake(Shiba Inu, 100)".-> p[StakingProxy];
  p[StakingProxy] --"stake(Shiba Inu, 100)"--> w[World];
  p -- "transferFrom(Client,World,100)" --> e[Shiba Inu];
  w --forward call--> i[private StakingSystem]
  i --"BalanceTable.set(Client, 100)"--> w;
```

rather than:

```mermaid
flowchart TB;
  C[client] -."stake(Shiba Inu, 100)".-> w[World];
  w --forward call--> i[public StakingSystem]
  i -- "transferFrom(Client,World,100)" --> e[Shiba Inu];
  i <--"TokenAddress.get(Shiba Inu)"--> w;
  i --"BalanceTable.set(Client, 100)"--> w;
```

The latter works in diamond pattern and mud1, but only works in mud2 if `StakingSystem` is limited to a single world.

Please let me know if the former is how mud2 is intended to be used. It took a lot of back and forth to get to this point 😅. Thanks again for patiently explaining everything @dk1a @alvrs!

**Comment by Boffee** [comment]
Actually, the architecture in the first chart doesn't work because staking not be accessible from the World. We would need to do something like:

```mermaid
flowchart TB;
  C[client] -."1. stake(Shiba Inu, 100)".-> p[StakingProxy];
  p[StakingProxy] --"2. stake(Shiba Inu, 100)"--> w[World];
  w --3. forward call--> i[private StakingSystem]
  i --"4. BalanceTable.set(Client, 100)"--> w;
  i --5. system hook--> w;
  w --6. system hook--> p;
  p -- "7. transferFrom(Client,World,100)" --> e[Shiba Inu];
```

so that we can have a world entry point:

```mermaid
flowchart TB;
  C[client] -."1. stake(Shiba Inu, 100)".-> w[World];
  w --2. forward call--> pi[public StakingSystem];
  pi --3. delegatecall?--> i[private StakingSystem]
  i --"4. BalanceTable.set(Client, 100)"--> w;
  i --5. system hook--> w;
  w --6. system hook--> p;
  p[StakingProxy] -- "7. transferFrom(Client,World,100)" --> e[Shiba Inu];
```

**Comment by holic** [comment]
(wow, this is the first time I am seeing these charts, very cool)
